src/echo/echo.rs

```diff
@@ -36,119 +36,90 @@ static HELP: &str = r#"
  \\xHH    byte with hexadecimal value HH (1 to 2 digits)
 "#;
 
-enum Base {
-    B8 = 8,
-    B16 = 16,
-}
-
-struct Opts {
-    newline: bool,
-    escape: bool,
+fn parse_code(input: &mut Peekable<Chars>, base: u32, max_digits: u32, bits_per_digit: u32) -> Option<char> {
+    let mut ret = 0x80000000;
+    for _ in 0..max_digits {
+        match input.peek().and_then(|c| c.to_digit(base)) {
+            Some(n) => ret = (ret << bits_per_digit) | n,
+            None => break,
+        }
+        input.next();
+    }
+    std::char::from_u32(ret)
 }
 
-fn convert_str(string: &[u8], index: usize, base: Base) -> (char, usize) {
-    let (max_digits, is_legal_digit): (usize, fn(u8) -> bool) = match base {
-        Base::B8 => (3, |c| (c as char).is_digit(8)),
-        Base::B16 => (2, |c| (c as char).is_digit(16)),
-    };
-
-    let mut bytes = vec![];
-    for offset in 0..max_digits {
-        if string.len() <= index + offset as usize {
-            break;
-        }
-        let c = string[index + offset as usize];
-        if is_legal_digit(c) {
-            bytes.push(c as u8);
-        } else {
-            break;
+pub fn unescape(input: &str) -> String {
+    let mut output = String::with_capacity(input.len());
+    let mut iter = input.chars().peekable();
+    while let Some(c) = iter.next() {
+        match c {
+            '\\' => match iter.next() {
+                Some(c) => match c {
+                    '\\' => output.push('\\'),
+                    'a' => output.push('\x07'),
+                    'b' => output.push('\x08'),
+                    'c' => break,
+                    'e' => output.push('\x1b'),
+                    'f' => output.push('\x0c'),
+                    'n' => output.push('\n'),
+                    'r' => output.push('\r'),
+                    't' => output.push('\t'),
+                    'v' => output.push('\x0b'),
+                    'x' => match parse_code(&mut iter, 16, 2, 4) {
+                        Some(c) => output.push(c),
+                        None => output.push_str("\\x"),
```

Comment:
I think you can avoid all the `output.push()` stuff everywhere if you do something like:
```rust
let ch = if c == '\\' {
    match iter.next() {
        // ...
        'v' => '\x0b',
        'x' => parse_code(&mut iter, 16, 2, 4).unwrap_or_else(|| {
            output.push('\\');
            'x'
        }),
        // ...
    }
} else {
    c
};
output.push(ch);
```

You can also avoid the `else` entirely if you make `c` mutable.